File: Javascript-DSA/Arrays/AlgoExpert/Two-Pointers-Approach/threeNumberSum.js
//? https://www.algoexpert.io/questions/Three%20Number%20Sum
//! O(n^2) | space O(n)
function threeNumberSum(array, targetSum) {
    array.sort( (a, b) => a - b);
    const triplets = [];

    for (let i = 0; i < array.length - 2; i++) {
        let left = i + 1;
        let right = array.length - 1;
        while (left < right) {
            const currentSum = array[i] + array[left] + array[right];

            if (currentSum === targetSum) {
                triplets.push([array[i], array[left], array[right]]);
                left++; right--;
            }
            else if (currentSum < targetSum) {
                    left++;
            }
            else if (currentSum > targetSum) {
                right--;
            }
        }
    }
    return triplets;
}

console.log(threeNumberSum(array, targetSum));

let array = [1, 4, 4, 5, 5, 5, 6, 6 ,11];
array.sort( (a, b) => a - b)
let targetSum = 10;
let result = [];


/** 
//! O(n^3) | space O(n)

for(let i = 0; i <= array.length - 1; i++) {
    let a = array[i];   
    for(let j = i + 1; j <= array.length - 2; j++) {
        let b = array[j];
        for(let k = j + 1; k <= array.length - 1; k++) {
            let c = array[k];
            let d = a + b + c;
            if(d === targetSum) {
                result.push([a, b, c]); 
            }
        }
    }
}

for(let i = 0; i < result.length; i++) {
    result[i].sort( (a, b) => a - b);
}

console.log(result);

*/

File: Hash Table/three_number_sum.java
import java.util.*;

/**
 *

 Write a function that takes in a non-empty array of distinct integers and an integer representing a target sum. The function should find all triplets in the array that sum up to the target sum and return a two-dimensional array of all these triplets. The numbers in each triplet should be ordered in ascending order, and the triplets themselves should be ordered in ascending order with respect to the numbers they hold.

 If no three numbers sum up to the target sum, the function should return an empty array.
 Sample Input

 array = [12, 3, 1, 2, -6, 5, -8, 6]
 targetSum = 0

 Sample Output

 [[-8, 2, 6], [-8, 3, 5], [-6, 1, 5]]


 */
public class ThreeNumberSum {
    // O(n^2) time | O(n) space
    public static void main(String[] args){
        int[] array = {12, 3, 1, 2, -6, 5, -8, 6};
        int targetSum = 0;
        List<Integer[]> triplets = solve(array, targetSum);
        for (var triplet: triplets) System.out.println(Arrays.toString(triplet));
    }
    public static List<Integer[]> solve(int[] array, int targetSum) {
        Arrays.sort(array);
        List<Integer[]> triplets = new ArrayList<>();

        for(int i = 0; i < array.length - 2; i++) {
            int left = i + 1;
            int right = array.length - 1;
            while(left < right) {
                int currentSum = array[i] + array[left] + array[right];
                if(currentSum == targetSum) {
                    Integer[] newTriplet = {array[i], array[left], array[right]};
                    triplets.add(newTriplet);
                    left++;
                    right--;
                } else if(currentSum < targetSum) left++;
                else right--;
            }
        }
        return triplets;
    }
}